pkg/envoy/lds/ingress.go

```diff
@@ -57,27 +65,43 @@ func newIngressFilterChain(cfg configurator.Configurator, svc service.MeshServic
 	}
 }
 
-func getIngressFilterChains(svc service.MeshService, cfg configurator.Configurator) []*xds_listener.FilterChain {
+func (lb *listenerBuilder) getIngressFilterChains(svc service.MeshService) []*xds_listener.FilterChain {
 	var ingressFilterChains []*xds_listener.FilterChain
 
-	if cfg.UseHTTPSIngress() {
-		// Filter chain with SNI matching enabled for HTTPS clients that set the SNI
-		ingressFilterChainWithSNI := newIngressFilterChain(cfg, svc)
-		ingressFilterChainWithSNI.Name = inboundIngressHTTPSFilterChain
-		ingressFilterChainWithSNI.FilterChainMatch.ServerNames = []string{svc.ServerName()}
-		ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithSNI)
+	protocolToPortMap, err := lb.meshCatalog.GetPortToProtocolMappingForService(svc)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error retrieving port to protocol mapping for service %s", svc)
+		return ingressFilterChains
 	}
 
-	// Filter chain without SNI matching enabled for HTTP clients and HTTPS clients that don't set the SNI
-	ingressFilterChainWithoutSNI := newIngressFilterChain(cfg, svc)
-	ingressFilterChainWithoutSNI.Name = inboundIngressHTTPFilterChain
-	ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithoutSNI)
+	// Create protocol specific inbound filter chains per port to handle different ports serving different protocols
+	for port, appProtocol := range protocolToPortMap {
+		switch appProtocol {
+		case httpAppProtocol:
+			// Ingress filter chain for HTTP port
+			if lb.cfg.UseHTTPSIngress() {
+				// Filter chain with SNI matching enabled for HTTPS clients that set the SNI
+				ingressFilterChainWithSNI := newIngressHTTPFilterChain(lb.cfg, svc, port)
+				ingressFilterChainWithSNI.Name = fmt.Sprintf("%s:%d", inboundIngressHTTPSFilterChain, port)
+				ingressFilterChainWithSNI.FilterChainMatch.ServerNames = []string{svc.ServerName()}
+				ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithSNI)
+			}
+
+			// Filter chain without SNI matching enabled for HTTP clients and HTTPS clients that don't set the SNI
+			ingressFilterChainWithoutSNI := newIngressHTTPFilterChain(lb.cfg, svc, port)
+			ingressFilterChainWithoutSNI.Name = fmt.Sprintf("%s:%d", inboundIngressNonSNIFilterChain, port)
+			ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithoutSNI)
```

Comment:
Wondering if you could benefit of a deepcopy (assuming reflect pkg will clone and repoint the cloned structs) of the result instead of calling `newIngressHTTPFilterChain` two times since I don't see difference in the invocations, hence I'd expect the results to be the same, avoiding k8s cache accesses if there were any.

Not a must-change though.

Reply:
Good point, I haven't changed the existing behavior in this change. Eventually I foresee a use case to just parameterize the difference in the 2 chains and let the function build it as expected.

Reply:
makes sense, I sort of imagined we would want to keep it for such